Support replication connections through PgBouncer

JelteF opened the pull request:
In session pooling mode PgBouncer is pretty much a transparent proxy,
i.e. the client does normally not even need to know that PgBouncer is in
the middle. This allows things like load balancing and failovers without
the client needing to know about this at all. But as soon as replication
connections are needed, this was not possible anymore, because PgBouncer
would reject those instead of proxying them to the right server.

This PR fixes that by also proxying replication connections. They are
handled pretty differently from normal connections though. A client and
server replication connection will form a strong pair, as soon as one is
closed the other is closed too. So, there's no caching of the server
replication connections, like is done for regular connections. Reusing
replication connections comes with a ton of gotchas. Postgres will
throw errors in many cases when trying to do so. So simply not
doing it seems like a good tradeoff for ease of implementation.
Especially because replication connections are pretty much always
very long lived. So re-using them gains pretty much no performance
benefits.

Fixes #382

Depends on #945

Melkij commented:
>  So re-using them gains pretty much no performance benefits.

Moreover, replication connections cannot be reliably reused at all. Attempting to do so will lead to various mysteries like
```
elog(ERROR, "no record found"); /* shouldn't happen */
```
in wal sender code. Wal sender implicitly assumes that it works with one wal receiver.
(Based on my experience with a single cloud service where admins tried to use connection pooling for replication too)

JelteF commented:
> Moreover, replication connections cannot be reliably reused at all.

Thanks. I updated the wording in the PR to include that.

JelteF commented on src/admin.c:
```diff
@@ -882,12 +890,15 @@ static bool admin_show_pools(PgSocket *admin, const char *arg)
 	statlist_for_each(item, &pool_list) {
 		pool = container_of(item, PgPool, head);
 		waiter = first_socket(&pool->waiting_client_list);
+		/* XXX: should maxwait include replication connections? */
```
Conceptually it would make sense to include replication connections too, but I also don't think it's super important.

petere commented on doc/config.md:
```diff
@@ -1249,7 +1249,7 @@ The file follows the format of the PostgreSQL `pg_hba.conf` file
 
 * Supported record types: `local`, `host`, `hostssl`, `hostnossl`.
-* Database field: Supports `all`, `sameuser`, `@file`, multiple names.  Not supported: `replication`, `samerole`, `samegroup`.
+* Database field: Supports `all`, `replication` `sameuser`, `@file`, multiple names.  Not supported: `samerole`, `samegroup`.
```
comma

petere commented on include/bouncer.h:
```diff
@@ -490,6 +507,12 @@ struct PgDatabase {
 };
 
+enum ReplicationType {
+	REPLICATION_NONE,
```
Maybe set this one to `= 0`, so this can be used as a boolean.

petere commented on include/bouncer.h:
```diff
@@ -54,6 +54,8 @@ enum SocketState {
 	CL_WAITING,		/* pool->waiting_client_list */
 	CL_WAITING_LOGIN,	/*   - but return to CL_LOGIN instead of CL_ACTIVE */
+	CL_WAITING_LOGICAL_REP,	/* pool->waiting_logical_replication_list */
+	CL_WAITING_PHYSICAL_REP,	/* pool->waiting_physical_replication_list */
```
Why do we need separate lists for logical and physical replication?

JelteF replied:
I guess you're right and we probably don't need that (after changing a few things).

JelteF commented:
I fixed all the things I wanted to in this PR. So I think it's ready for a more thorough review now.

petere commented on .editorconfig:
```diff
@@ -8,7 +8,7 @@ insert_final_newline = true
 trim_trailing_whitespace = true
 
-[*.{c,h}]
+[*.{c,h,eval,rules}]
```
This had me confused. What is *.eval? Maybe list the full file names here. And make it a separate stanza from the C-code related things.

petere commented on include/common/postgres_compat.h:
```diff
@@ -30,6 +32,8 @@
 #define pg_sha256_final(ctx, dst) sha256_final(ctx, dst)
 
+extern bool parse_bool(const char *value, bool *result);
+extern bool parse_bool_with_len(const char *value, size_t len, bool *result);
```
Isn't this redundant with `include/common/builtins.h`?

petere commented on include/bouncer.h:
```diff
@@ -287,6 +289,13 @@ struct PgPool {
+	 * available. These requests are waiting for a new server connection to be
+	 * opened, before the request can be forwarded.
+	 */
+	struct StatList waiting_replication_client_list;
```
Why is there a separate list for this? Why aren't these just part of the normal waiting_client_list? Maybe there is a reason, but it could be explained here. I mean, there is no separate list of active replication clients, so why a separate list of waiting replication clients?

JelteF replied:
Good question. I added the reason to the comment now (also to the one for cancel connections since it's the same reason).

JelteF replied:
I actually removed the state completely now. It turned out not to be necessary (given some other changes).

petere commented on test/hba_test.rules:
```diff
@@ -5,10 +5,6 @@
-# ws
- 	 	 	
-			# z
-
```
Wasn't this here for some parsing test?

JelteF replied:
I couldn't gather that from the commit message that added the file 8 years ago. The lines seemed like nonsense to me and they were causing diff when I saved because they had trailing whitespace that was removed by the editorconfig settings.

I added them back now and disabled the trailing whitespace removal for this file.

eulerto commented on test/test_replication.py:
```diff
@@ -0,0 +1,250 @@
+
+
+@pytest.mark.skipif(
+    "PG_MAJOR_VERSION < 10", reason="pg_basebackup on PG9.6 fails with a weird error"
```
I'm curious about the "weird error". Is it a random failure? Or it is because of a behavior change (it streams WAL by default)?

JelteF replied:
I took a bit closer look and I was able to fix it by the flag `-X stream`.

eulerto reviewed (commented):
This is _not_ a complete review. I want to test some cases to see if I can break it. One suggestion is to add a test to exercise the `waiting_replication_client_list`.

JelteF commented on test/test_replication.py:
```diff
@@ -0,0 +1,250 @@
+    # its three or more.
+    assert (
+        bouncer.sql_value("SELECT count(*) FROM test_logical_rep_subscriber.table") >= 3
+    )
```
Thanks for the suggestion to use different databases for the source and target. I agree that makes the test much easier to understand.

JelteF commented on src/server.c:
```diff
@@ -479,12 +480,12 @@ static bool handle_connect(PgSocket *server)
 		/* proceed with login */
 		if (server_connect_sslmode > SSLMODE_DISABLED && !is_unix) {
-			slog_noise(server, "P: SSL request");
+			slog_debug(server, "P: SSL request");
```
Reverted this change.

JelteF commented on src/server.c:
```diff
@@ -619,7 +620,9 @@ bool server_proto(SBuf *sbuf, SBufEvent evtype, struct MBuf *data)
-		if (pool_pool_mode(pool) != POOL_SESSION || server->state == SV_TESTED || server->resetting) {
+		if (server->state == SV_TESTED || server->resetting || (
+			    pool_pool_mode(pool) != POOL_SESSION && !server->replication
+			    )) {
```
Removed this change. I think I added it for debugging purposes and it's not actually necessary.

JelteF commented on include/bouncer.h:
```diff
@@ -301,10 +303,30 @@ struct PgPool {
+	 * This is a separate list from waiting_client_list, because we need to trigger
+	 * opening a completely new server connection for them. The clients in
+	 * waiting_client_list can also use a currently idle server connection.
+	 *
```
I rephrased it, I hope it's clearer now.

JelteF replied:
Actually... I deleted the CL_WAITING_REPLICATION state completely now. I did still update the description of the wait_cancel_req_list to hopefully make it clearer.

eulerto commented:
Isn't the cancel key change a requirement to this feature? I suggest you to create a separate commit for it.

eulerto commented:
> Isn't the cancel key change a requirement to this feature? I suggest you to create a separate commit for it.

Ops... I realized that you did it. Another suggestion is to explain in the commit message that it is a requirement for this feature.

JelteF commented:
> Ops... I realized that you did it. Another suggestion is to explain in the commit message that it is a requirement for this feature.

I added a commit message and moved it before the actual replication change, but while doing that I realized I thought the change was special enough to warrant its own PR, so I made a new one: https://github.com/pgbouncer/pgbouncer/pull/945

petere commented on .editorconfig:
I'm not sure these changes are necessary. Can we just leave it alone?

JelteF replied:
I extracted this and moved it to: https://github.com/pgbouncer/pgbouncer/pull/1017

petere commented on include/common/postgres_compat.h:
```diff
@@ -30,6 +32,5 @@
 #define pg_sha256_update(ctx, data, len) sha256_update(ctx, data, len)
 #define pg_sha256_final(ctx, dst) sha256_final(ctx, dst)
 
-
```
Let's keep that.

petere commented on include/common/postgres_compat.h:
```diff
@@ -15,6 +16,7 @@
 
 #define lengthof(array) (sizeof (array) / sizeof ((array)[0]))
 #define pg_hton32(x) htobe32(x)
+#define pg_strncasecmp strncasecmp
```
These aren't necessarily the same thing. `pg_strncasecmp()` only case-folds ASCII characters. Maybe that's ok for current use? But we should be careful about documenting this.

petere commented on src/hba.c:
```diff
@@ -31,8 +31,9 @@ enum RuleType {
-#define NAME_SAMEUSER   2
+#define NAME_ALL                1
+#define NAME_SAMEUSER           2
+#define NAME_REPLICATION        4
```
We should comment here that this is a bit mask. It wasn't clear from the first two values, but with the third one it becomes relevant.

petere commented on src/proto.c:
```diff
@@ -584,15 +593,69 @@ bool answer_authreq(PgSocket *server, PktHdr *pkt)
+		 * the varcache parameters to the startup message allows us to
+		 * skip the dance that involves sending Query packets and
+		 * waiting for responses.
+		 */
```
Could we also apply this to normal connections, to get the same optimization benefit?

JelteF replied:
Then we would need to link normal connections here already too. We don't do that currently. I'd expect that to cause some problems if we did that. It might be possible, but I don't think it fits well in this PR to do so.

petere commented on src/server.c:
```diff
@@ -122,7 +122,18 @@ static bool handle_server_startup(PgSocket *server, PktHdr *pkt)
+		 * However, we don't do this if it's a replication connection,
+		 * because those can fail due to a variety of missing
+		 * permissions, while normal connections would still be able
+		 * connect and query the database just fine.
```
I don't understand this distinction. Can you give examples?

petere commented on include/bouncer.h:
```diff
@@ -305,6 +306,11 @@ struct PgPool {
 	 * These requests are waiting for a new server connection to be opened,
 	 * before the request can be forwarded.
+	 *
+	 * This is a separate list from waiting_client_list, because we want to
```
Is this related to this patch?

JelteF replied:
I extracted this small change and moved it to: https://github.com/pgbouncer/pgbouncer/pull/1016

petere commented on src/server.c:
```diff
@@ -192,6 +203,18 @@ static bool handle_server_startup(PgSocket *server, PktHdr *pkt)
+		return POOL_SESSION;
+	return pool_pool_mode(server->pool);
+}
+
```
I'm worried that if we have both `server_pool_mode()` and still the old `pool_pool_mode()`, it will be difficult to enforce that the right one is called in each situation.

I wonder if we could alleviate that by allocating a `PgPool` structure for each replication connection, but only allow one connection through it. Then we can keep using `pool_pool_mode()`, and maybe also have fewer changes elsewhere.  (For example, this patch doesn't appear to have anything for statistics or `SHOW` commands. I wonder if this still works sensibly for connections that are not associated with pools. And of course, we do want some sort of monitoring support in any case.)

JelteF replied:
It's a reasonable worry, because I could definitely see that be an issue. To avoid the issue I added a warning to the function comment of `pool_pool_mode()` and renamed it to `probably_wrong_pool_pool_mode()`, because it should basically never be used directly (the only place where it should be used is in `SHOW pools`).

I did that instead of your proposal, because I do want to keep the current general design. The reason for that is that this is more in line with the future plan I have of allowing a client to choose if they want transaction or session pooling: https://github.com/pgbouncer/pgbouncer/issues/997 Basically for that near-future plan the pooling mode would be a property of the connection, not of the pool. Just like it is a property of the connection now for replication connections. 

Regarding statistics, I think it would make sense to add a replication and pool_mode column to `show servers/clients`. But the existing reporting also works for replication connections.

JelteF replied:
FYI I added a replication column to `SHOW CLIENTS/SERVERS`.  I think that should be enough info for debugging any issues.

JelteF commented on src/janitor.c:
```diff
@@ -534,13 +541,13 @@ static void pool_server_maint(PgPool *pool)
 	check_unused_servers(pool, &pool->idle_server_list, 1);
 
 	/* disconnect close_needed active servers if server_fast_close is set */
```
done
